### DISCUSSION TITLE
Avoiding indirect linkage issues

### DISCUSSION THREAD

**Discussion by mahrud** [Homebrew/discussions#5368] [Writing Formulae/Casks]
### Description of issue

I have a formula on a tap that builds a rather large project using CMake. I'm wondering if anyone with experience using CMake and Homebrew could help me fix the following indirect linkage issue, or point me to other projects that use CMake and depend on `libomp`:
```
GitHub Actions / test-bot (macos-14)
`brew linkage --cached --test --strict macaulay2/tap/macaulay2` failed on macOS Sonoma (14) on Apple Silicon!
Indirect dependencies with linkage:
  icu4c
  libomp
```
```
GitHub Actions / test-bot (ubuntu-latest)
`brew linkage --cached --test --strict macaulay2/tap/macaulay2` failed on Linux!
Indirect dependencies with linkage:
  gcc
```
(The specific pull request that resulted on this is [here](https://github.com/Macaulay2/homebrew-tap/pull/223/checks).)

Surely there's a way to statically link libomp, icu4c, and gcc, but I can't figure out how.

**Comment by SMillerDev** [comment]
> Surely there's a way to statically link libomp, icu4c, and gcc, but I can't figure out how.

Very few projects do static linking in Homebrew since it's actively discouraged. Your usage of `std_cmake_args` almost certainly means that it's using a dynamic linking build.

**Reply by mahrud** [reply]
Good point, but then what is the warning asking me to do?

Also, `icu4c` is particularly annoying because it's linking with a shared library in a cellar:
```
	libicudata.so.74 => /home/linuxbrew/.linuxbrew/Cellar/icu4c/74.2/lib/libicudata.so.74 (0x00007fe4e2000000)
```
Which causes breakages whenever it's updated.

**Reply by SMillerDev** [reply]
> Good point, but then what is the warning asking me to do?

To declare the dependencies you have instead of hoping they will always be available with your current set of dependencies.

> Also, icu4c is particularly annoying because it's linking with a shared library in a cellar.

It's also a versioned library so it'll almost always need a version bump when it's updated

**Comment by ipatch** [comment]
i came across this same error / warning with the homebrew-freecad tap this afternoon (snippets & links below). however the build process from test-bot still seemed to continue and upload a bottle'd artifact. so not sure if these are just warnings or if it will cause the app to crash at runtime, as i haven't tested the bottle'd artifact manually but the tests via the ci seemed to pass.

interestingly enough the _indirect linkage_ is described as both a _warning_ and _failure_ so not sure of the severity here 🤷‍♂️

```
==> Testable dependents:

Warning: 1 failed step ignored!
brew linkage --cached --test --strict freecad/freecad/pyside2@5.15.11_py310
```

https://github.com/FreeCAD/homebrew-freecad/actions/runs/9177500809/job/25235238110?pr=512#step:18:70

snippet grabbed from the above link.

```
==> brew linkage --test freecad/freecad/pyside2@5.15.11_py310
==> brew linkage --cached --test --strict freecad/freecad/pyside2@5.15.11_py310
==> FAILED
Full linkage --cached --test --strict freecad/freecad/pyside2@5.15.11_py310 output
  Indirect dependencies with linkage:
    llvm
Truncated linkage --cached --test --strict freecad/freecad/pyside2@5.15.11_py310 output
  Warning: Indirect dependencies with linkage:
    llvm
  
==> brew linkage --cached freecad/freecad/pyside2@5.15.11_py310
==> brew install --only-dependencies --include-test freecad/freecad/pyside2@5.15.11_py310
==> brew test --verbose freecad/freecad/pyside2@5.15.11_py310
```

---

after your ci run went through test-bot step. did the workflow / run still manage to upload a bottle'd artifact for your repo / tap?

**Reply by Bo98** [reply]
> so not sure if these are just warnings or if it will cause the app to crash at runtime

They won't cause any issues at runtime, but things could break if the dependency up the chain stops using `llvm` without you knowing.

They are just warnings for now but we might make them errors in the future so that we can optimise the test-bot dependent testing. If everything is declared directly, there won't be as much need for recursive dependent testing.

We generally recommend declaring dependencies directly anyway, particularly if you are depending on homebrew-core formulae. The dependencies of a homebrew-core formula may change without notice so you shouldn't assume that the indirect dependency will always be around.

An alternative to declaring dependencies directly could be configuring your formula to avoid the opportunistic linkage altogether, should you feel the dependency isn't really necessary for your use case.

> is described as both a warning and failure 

This is just a weirdness from the way test-bot reports issues. The summary at the bottom is correct - it's a warning. But we recommend you declare dependencies directly.

**Reply by ipatch** [reply]
well unfortunately with my use case the indirect linkage was to the upstream `llvm` formula, but if i declare llvm (latest version, ie. llvm.rb) directly in the `pyside2@5.15.11_py310.rb` formula file then the build will fail. for pyside2 i have to use `llvm@15` as i believe homebrew-core also uses `llvm@15` for pyside2. i haven't checked ci run logs for homebrew-core to see if the _indirect linkage_ warning is present as well, but it seems there could be some edge cases that would crop up and be problematic if this warning became and error and prevented the building of bottle'd artifacts. _just my 2 cents though_.

**Reply by Bo98** [reply]
If Homebrew is reporting it then `otool`/`ldd` is likely reporting that `pyside2` is actually using `llvm` without you knowing.

If you do see it being incorrect let me know and I'll have a look. I'll also have a look at homebrew-core's pyside2.

**Reply by mahrud** [reply]
In my case, I know what things from gcc and icu4c are being used, but they are so tiny and incidental that it would be _significantly_ more convenient if those small pieces were statically linked so that users didn't need to download gcc. Also, since it's a self-managed tap, rebuilding bottles every time icu4c updates has been a big hassle.

**Reply by Bo98** [reply]
Nothing has changed in Homebrew in that regards. Your formula was always indirectly linking to icu4c - it just wasn't telling you before.

How exactly that is addressed totally depends on your build script. I suspect `Boost::regex`'s CMake script is inserting this in: https://github.com/boostorg/boost_install/issues/47. You might be add a `STATIC IMPORTED` `icudata` CMake target etc that points to the static library. The default behaviour otherwise will be `-licudata`.

(The gcc one on Linux is a Homebrew bug however that has already been fixed.)

**Reply by PatTheMav** [reply]
I've ran across the same warnings just now (had not updated my custom taps in some time) and in this specific case it's FFmpeg which explicitly depends on `libvorbis` and `theora`, both of which depend on `libogg` which is somehow added as a direct dependency to FFmpeg itself.

Knowing FFmpeg I would not be surprised if this is the result of `pkg-config` information and the output for `theora` indeed includes `ogg`:

`-L/opt/homebrew/Cellar/theora/1.1.1/lib -L/opt/homebrew/Cellar/libogg/1.3.5/lib -ltheora -logg`

So is this something that I can fix as a formula author or this something that needs to be fixed in the `theora` formula (or, as a third possibility, I misunderstand the indirect linking issue)?

To wit: The Linux build of FFmpeg complains about `libx11`, `libxcb`, and `libxext` being linked indirectly.

**Comment by ipatch** [comment]
for reference / posterity purposes, snippet copied from the raw logs.

https://github.com/Macaulay2/homebrew-tap/actions/runs/9087675907/job/24975839771#step:8:42397

```
2024-05-15T01:21:14.4744247Z Emacs Lisp files have been installed to:
2024-05-15T01:21:14.4745444Z   /home/linuxbrew/.linuxbrew/share/emacs/site-lisp/macaulay2
2024-05-15T01:21:14.4746497Z [34m==>[0m [1mSummary[0m
2024-05-15T01:21:14.8198829Z /home/linuxbrew/.linuxbrew/Cellar/macaulay2/1.24.05: 21,017 files, 520MB
2024-05-15T01:21:14.8212869Z [34m==>[0m [1mRunning `brew cleanup macaulay2`...[0m
2024-05-15T01:21:14.9134376Z [34m==>[0m [1mbrew linkage --test macaulay2/tap/macaulay2[0m
2024-05-15T01:21:16.0712854Z 
2024-05-15T01:21:16.0714300Z [34m==>[0m [1mbrew linkage --cached --test --strict macaulay2/tap/macaulay2[0m
2024-05-15T01:21:17.2205199Z Indirect dependencies with linkage:
2024-05-15T01:21:17.2205902Z   gcc
2024-05-15T01:21:17.2455802Z [31m==>[0m [1m[31mFAILED[0m[0m
2024-05-15T01:21:17.2541769Z ##[group]Truncated linkage --cached --test --strict macaulay2/tap/macaulay2 output
2024-05-15T01:21:17.2553798Z ##[warning]Indirect dependencies with linkage:
  gcc
  ```
